lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2523,59 +2549,107 @@ bool EliminateSliceConcat::run(Function *F, const CompilationContext &cctx) {
     // 2) merging Slices from different sources
     // e.g. Slice(A1)-Slice(B1)-Slice(A2)-Slice(B2), A1 and A2, B1 and B2 are
     // consecutive respectively
-    std::vector<std::vector<SliceNode *>> consecutiveSlices;
+    //
+    // Store consecutive slices along *any* dimension. If the consecutive
+    // slices' dimension is the same as the concat, the concat can be removed.
+    // If the slices' dimension is different from the concat, the nodes
+    // can be replaced with a single slice+reshape.
+    std::vector<std::pair<unsigned_t /* dimension of slicing */,
+                          std::vector<SliceNode *>>> consecutiveSlices;
     std::vector<SliceNode *> currConsecutiveSlices;
     SliceNode *lastSN = nullptr;
+    unsigned_t lastDim = -1;
```

Comment:
Since it's unsigned, the default value will be `2**32-1`. I hope that's not a valid dimension I would have to consider.